src/pybind/mgr/cephadm/http_server.py

```diff
@@ -37,17 +39,44 @@ def configure_cherrypy(self) -> None:
             'engine.autoreload.on': False,
         })
 
-    def run(self) -> None:
+    def configure(self) -> None:
+        self.configure_cherrypy()
+        self.agent.configure()
+        self.service_discovery.configure(self.mgr.service_discovery_port, self.mgr.get_mgr_ip())
+
+    def port_in_use(self, addr: str, port: int) -> bool:
         try:
-            self.configure_cherrypy()
-            self.agent.configure()
-            self.service_discovery.configure()
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sock.bind((addr, port))
+            sock.close()
+            return False
+        except socket.error:
+            return True
+
+    def config_update(self) -> None:
+        if self.service_discovery_port != self.mgr.service_discovery_port:
+            if (self.port_in_use(self.mgr.get_mgr_ip(), self.mgr.service_discovery_port)):
+                self.mgr.log.error(f'configured service discovery port {self.mgr.service_discovery_port} is already in use. '
+                                   f'Listening on old port {self.service_discovery_port}.')
+                return
+            else:
+                self.mgr.log.debug(f'Changing service discovery port from {self.service_discovery_port} to {self.mgr.service_discovery_port}...')
+                self.service_discovery_port = self.mgr.service_discovery_port
+                self.restart()
+
+    def restart(self) -> None:
+        cherrypy.engine.stop()
+        cherrypy.server.httpserver = None
+        self.configure()
+        cherrypy.engine.start()
```

Comment:
@epuertat during my testing I didn't see any suspicious behavior. At the begining I ran in some issues because when trying to restart the server (and bind the same port) I was getting port in use error because the actually the server didn't free the port. I add to add this callback in order for force calling the `unsbuscribe()` when stopping the server:

```
    def stop(self) -> None:
        # we must call unsubscribe before stopping the server,
        # otherwise the port is not released and we will get
        # an exception when trying to restart it
        self.unsubscribe()
        super().stop()
```
basides, in a previour PR I also found that setting the cherrypy server to None as in the following code helped to free stuff and re-create the server:

```
    def restart(self) -> None:
        cherrypy.engine.stop()
        cherrypy.server.httpserver = None
        self.configure()
        cherrypy.engine.start()
```
I found this _fix_ on SO [restart cherrypy](https://stackoverflow.com/questions/7254845/change-cherrypy-port-and-restart-web-server) and as far as I remember it helped me stop getting errors when trying to restart the server.